File: r/palettecore/R/metrics.R
# Perceptual and accessibility metrics — mirrors palettecore/metrics.py.

#' CIEDE2000 colour difference between two sRGB colours
#'
#' Computed via CIELAB under D65 (Sharma et al. 2005 formulation).
#'
#' @param rgb1,rgb2 Numeric vectors of length 3, sRGB channels in \[0, 1\].
#' @return CIEDE2000 distance (0 for identical colours, about 100 for black vs white).
#' @export
ciede2000 <- function(rgb1, rgb2) {
  lab1 <- srgb_to_cielab(rgb1)
  lab2 <- srgb_to_cielab(rgb2)
  L1 <- lab1[1]; a1 <- lab1[2]; b1 <- lab1[3]
  L2 <- lab2[1]; a2 <- lab2[2]; b2 <- lab2[3]

  C1 <- sqrt(a1^2 + b1^2)
  C2 <- sqrt(a2^2 + b2^2)
  C_bar <- (C1 + C2) / 2
  G <- 0.5 * (1 - sqrt(C_bar^7 / (C_bar^7 + 25^7)))
  a1p <- (1 + G) * a1
  a2p <- (1 + G) * a2
  C1p <- sqrt(a1p^2 + b1^2)
  C2p <- sqrt(a2p^2 + b2^2)

  hp <- function(a, b) {
    if (a == 0 && b == 0) return(0)
    (atan2(b, a) * 180 / pi) %% 360
  }
  h1p <- hp(a1p, b1)
  h2p <- hp(a2p, b2)

  dLp <- L2 - L1
  dCp <- C2p - C1p
  if (C1p * C2p == 0) {
    dhp <- 0
  } else {
    dd <- h2p - h1p
    if (dd > 180) dd <- dd - 360 else if (dd < -180) dd <- dd + 360
    dhp <- dd
  }
  dHp <- 2 * sqrt(C1p * C2p) * sin((dhp * pi / 180) / 2)

  Lp_bar <- (L1 + L2) / 2
  Cp_bar <- (C1p + C2p) / 2
  if (C1p * C2p == 0) {
    hp_bar <- h1p + h2p
  } else {
    dd <- abs(h1p - h2p)
    ss <- h1p + h2p
    if (dd <= 180) {
      hp_bar <- ss / 2
    } else if (ss < 360) {
      hp_bar <- (ss + 360) / 2
    } else {
      hp_bar <- (ss - 360) / 2
    }
  }

  rad <- function(x) x * pi / 180
  T_ <- 1 - 0.17 * cos(rad(hp_bar - 30)) + 0.24 * cos(rad(2 * hp_bar)) +
    0.32 * cos(rad(3 * hp_bar + 6)) - 0.20 * cos(rad(4 * hp_bar - 63))
  d_theta <- 30 * exp(-(((hp_bar - 275) / 25)^2))
  R_C <- 2 * sqrt(Cp_bar^7 / (Cp_bar^7 + 25^7))
  S_L <- 1 + 0.015 * (Lp_bar - 50)^2 / sqrt(20 + (Lp_bar - 50)^2)
  S_C <- 1 + 0.045 * Cp_bar
  S_H <- 1 + 0.015 * Cp_bar * T_
  R_T <- -sin(rad(2 * d_theta)) * R_C

  sqrt((dLp / S_L)^2 + (dCp / S_C)^2 + (dHp / S_H)^2 +
    R_T * (dCp / S_C) * (dHp / S_H))
}

relative_luminance <- function(rgb) {
  lin <- srgb_to_linear(rgb)
  0.2126 * lin[1] + 0.7152 * lin[2] + 0.0722 * lin[3]
}

#' WCAG 2.x contrast ratio between two sRGB colours
#'
#' @param rgb1,rgb2 Numeric vectors of length 3, sRGB channels in \[0, 1\].
#' @return Contrast ratio between 1 and 21.
#' @export
contrast_ratio <- function(rgb1, rgb2) {
  y1 <- relative_luminance(rgb1)
  y2 <- relative_luminance(rgb2)
  (max(y1, y2) + 0.05) / (min(y1, y2) + 0.05)
}

greyscale_values <- function(rgbs) {
  apply(rgbs, 1, relative_luminance)
}

is_monotonic <- function(values, tol = 1e-4) {
  d <- diff(values)
  all(d <= tol) || all(d >= -tol)
}

File: palettecore/metrics.py
"""Perceptual and accessibility metrics: CIEDE2000, WCAG contrast, greyscale."""

from __future__ import annotations

import numpy as np

from .convert import srgb_to_cielab, srgb_to_linear


def ciede2000(rgb1: np.ndarray, rgb2: np.ndarray) -> float:
    """CIEDE2000 colour difference between two sRGB colours (Sharma et al. 2005)."""
    L1, a1, b1 = srgb_to_cielab(rgb1)
    L2, a2, b2 = srgb_to_cielab(rgb2)

    C1 = np.hypot(a1, b1)
    C2 = np.hypot(a2, b2)
    C_bar = (C1 + C2) / 2
    G = 0.5 * (1 - np.sqrt(C_bar**7 / (C_bar**7 + 25.0**7)))
    a1p, a2p = (1 + G) * a1, (1 + G) * a2
    C1p, C2p = np.hypot(a1p, b1), np.hypot(a2p, b2)

    def _hp(a, b):
        if a == 0 and b == 0:
            return 0.0
        return np.degrees(np.arctan2(b, a)) % 360.0

    h1p, h2p = _hp(a1p, b1), _hp(a2p, b2)

    dLp = L2 - L1
    dCp = C2p - C1p
    if C1p * C2p == 0:
        dhp = 0.0
    else:
        d = h2p - h1p
        if d > 180:
            d -= 360
        elif d < -180:
            d += 360
        dhp = d
    dHp = 2 * np.sqrt(C1p * C2p) * np.sin(np.radians(dhp) / 2)

    Lp_bar = (L1 + L2) / 2
    Cp_bar = (C1p + C2p) / 2
    if C1p * C2p == 0:
        hp_bar = h1p + h2p
    else:
        d = abs(h1p - h2p)
        s = h1p + h2p
        if d <= 180:
            hp_bar = s / 2
        elif s < 360:
            hp_bar = (s + 360) / 2
        else:
            hp_bar = (s - 360) / 2

    T = (
        1
        - 0.17 * np.cos(np.radians(hp_bar - 30))
        + 0.24 * np.cos(np.radians(2 * hp_bar))
        + 0.32 * np.cos(np.radians(3 * hp_bar + 6))
        - 0.20 * np.cos(np.radians(4 * hp_bar - 63))
    )
    d_theta = 30 * np.exp(-(((hp_bar - 275) / 25) ** 2))
    R_C = 2 * np.sqrt(Cp_bar**7 / (Cp_bar**7 + 25.0**7))
    S_L = 1 + 0.015 * (Lp_bar - 50) ** 2 / np.sqrt(20 + (Lp_bar - 50) ** 2)
    S_C = 1 + 0.045 * Cp_bar
    S_H = 1 + 0.015 * Cp_bar * T
    R_T = -np.sin(np.radians(2 * d_theta)) * R_C

    return float(
        np.sqrt(
            (dLp / S_L) ** 2
            + (dCp / S_C) ** 2
            + (dHp / S_H) ** 2
            + R_T * (dCp / S_C) * (dHp / S_H)
        )
    )


def relative_luminance(rgb: np.ndarray) -> float:
    lin = srgb_to_linear(rgb)
    return float(0.2126 * lin[0] + 0.7152 * lin[1] + 0.0722 * lin[2])


def contrast_ratio(rgb1: np.ndarray, rgb2: np.ndarray) -> float:
    """WCAG 2.x contrast ratio, 1:1 to 21:1."""
    y1, y2 = relative_luminance(rgb1), relative_luminance(rgb2)
    hi, lo = max(y1, y2), min(y1, y2)
    return (hi + 0.05) / (lo + 0.05)


def greyscale_values(rgbs: np.ndarray) -> np.ndarray:
    """Relative luminance of each colour — what the palette becomes in print."""
    return np.array([relative_luminance(c) for c in np.atleast_2d(rgbs)])


def is_monotonic(values: np.ndarray, tol: float = 1e-4) -> bool:
    d = np.diff(np.asarray(values, dtype=float))
    return bool(np.all(d <= tol) or np.all(d >= -tol))
